server/src/main/java/org/elasticsearch/gateway/BaseGatewayShardAllocator.java

```diff
@@ -45,40 +45,38 @@
     protected final Logger logger = LogManager.getLogger(this.getClass());
 
     /**
-     * Allocate unassigned shards to nodes (if any) where valid copies of the shard already exist.
+     * Allocate an unassigned shard to nodes (if any) where valid copies of the shard already exist.
      * It is up to the individual implementations of {@link #makeAllocationDecision(ShardRouting, RoutingAllocation, Logger)}
      * to make decisions on assigning shards to nodes.
      *
      * @param allocation the allocation state container object
+     * @param shardRouting the shard to allocate
+     * @param unassignedAllocationHandler handles the allocation of the current shard
      */
-    public void allocateUnassigned(RoutingAllocation allocation) {
-        final RoutingNodes routingNodes = allocation.routingNodes();
-        final RoutingNodes.UnassignedShards.UnassignedIterator unassignedIterator = routingNodes.unassigned().iterator();
-        while (unassignedIterator.hasNext()) {
-            final ShardRouting shard = unassignedIterator.next();
-            final AllocateUnassignedDecision allocateUnassignedDecision = makeAllocationDecision(shard, allocation, logger);
+    public void allocateUnassigned(RoutingAllocation allocation, ShardRouting shardRouting,
```

Comment:
nit: swap argument order to make it the same as in `ExistingShardsAllocator`:
```suggestion
    public void allocateUnassigned(ShardRouting shardRouting, RoutingAllocation allocation,
```